File: mongoSchema/globals.ts
import mongoose, { Schema } from "mongoose";

export interface globalsInterface {
  BOT_CLIENT_ID: string;
  CASH_CONVERTERS: boolean;
  CASH_CONVERTERS_CHANNEL_ID: string | null;
  CASH_CONVERTERS_ROLE_ID: string | null;
  COMMAND_PERMISSION_ROLE_ID: string | null;
  CS_CHANNEL_ID: string | null;
  CS_ITEMS: boolean;
  CS_MARKET_CHANNEL_ID: string | null;
  CS_MARKET_ROLE_ID: string | null;
  CS_ROLE_ID: string | null;
  DISCORD_GUILD_ID: string;
  DISCORD_TOKEN: string;
  EBAY: boolean;
  EBAY_CHANNEL_ID: string | null;
  EBAY_ROLE_ID: string | null;
  ERROR_CHANNEL_ID: string | null;
  GUMTREE: boolean;
  GUMTREE_CHANNEL_ID: string | null;
  GUMTREE_ROLE_ID: string | null;
  SALVOS: boolean;
  SALVOS_CHANNEL_ID: string | null;
  SALVOS_ROLE_ID: string | null;
  STEAM_QUERY: boolean;
  STEAM_QUERY_CHANNEL_ID: string | null;
  STEAM_QUERY_ROLE_ID: string | null;
}

const GlobalsSchema = new Schema<globalsInterface>({
  BOT_CLIENT_ID: { type: String, required: true },
  CASH_CONVERTERS: { type: Boolean, required: true },
  CASH_CONVERTERS_CHANNEL_ID: { type: String, default: null },
  CASH_CONVERTERS_ROLE_ID: { type: String, default: null },
  COMMAND_PERMISSION_ROLE_ID: { type: String, default: null },
  CS_CHANNEL_ID: { type: String, default: null },
  CS_ITEMS: { type: Boolean, required: true },
  CS_MARKET_CHANNEL_ID: { type: String, default: null },
  CS_MARKET_ROLE_ID: { type: String, default: null },
  CS_ROLE_ID: { type: String, default: null },
  DISCORD_GUILD_ID: { type: String, required: true },
  DISCORD_TOKEN: { type: String, required: true },
  EBAY: { type: Boolean, required: true },
  EBAY_CHANNEL_ID: { type: String, default: null },
  EBAY_ROLE_ID: { type: String, default: null },
  ERROR_CHANNEL_ID: { type: String, default: null },
  GUMTREE: { type: Boolean, required: true },
  GUMTREE_CHANNEL_ID: { type: String, default: null },
  GUMTREE_ROLE_ID: { type: String, default: null },
  SALVOS: { type: Boolean, required: true },
  SALVOS_CHANNEL_ID: { type: String, default: null },
  SALVOS_ROLE_ID: { type: String, default: null },
  STEAM_QUERY: { type: Boolean, required: true },
  STEAM_QUERY_CHANNEL_ID: { type: String, default: null },
  STEAM_QUERY_ROLE_ID: { type: String, default: null },
});

const Globals = mongoose.model<globalsInterface>("Globals", GlobalsSchema);

export default Globals;

File: goserver/internal/db/query/globals.go
package query

import (
	"context"
	"dealscanner/internal/config"
	"dealscanner/internal/models"
	"errors"

	"gorm.io/gorm"
)

func (q *Query) GetGlobals(ctx context.Context) (*models.Globals, error) {
	if cached, ok := defaultGlobalsCache.get(); ok {
		return cached, nil
	}

	g, err := q.Globals.WithContext(ctx).First()
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil
		}
		return nil, err
	}
	defaultGlobalsCache.set(g)
	return g, nil
}

func (q *Query) UpdateGlobals(ctx context.Context, g *models.Globals) error {
	if err := q.Globals.WithContext(ctx).Save(g); err != nil {
		return err
	}
	defaultGlobalsCache.set(g)
	return nil
}

func (q *Query) SeedGlobals(ctx context.Context, cfg *config.Config) error {
	g, err := q.Globals.WithContext(ctx).First()
	if errors.Is(err, gorm.ErrRecordNotFound) {
		g = &models.Globals{
			ID:                   "1",
			CsTradeDollarRatio:   1.0,
			LootFarmDollarRatio:  1.0,
			CsDealsDollarRatio:   1.0,
			TradeitGgDollarRatio: 1.0,
		}
	} else if err != nil {
		return err
	}

	// The environment is the deployment source of truth for scanner switches,
	// Discord credentials, channels, and roles. Keep an existing migrated row
	// synchronized instead of retaining stale defaults indefinitely.
	g.BotClientId = cfg.BotClientId
	g.DiscordGuildId = cfg.DiscordGuildId
	g.DiscordToken = cfg.DiscordToken
	g.CashConverters = cfg.CashConverters
	g.CashConvertersChannelId = cfg.CashConvertersChannelId
	g.CashConvertersRoleId = cfg.CashConvertersRoleId
	g.CommandPermissionRoleId = cfg.CommandPermissionRoleId
	g.CsChannelId = cfg.CsChannelId
	g.CsItems = cfg.CsItems
	g.CsMarketChannelId = cfg.CsMarketChannelId
	g.CsMarketRoleId = cfg.CsMarketRoleId
	g.CsRoleId = cfg.CsRoleId
	g.Ebay = cfg.Ebay
	g.EbayChannelId = cfg.EbayChannelId
	g.EbayRoleId = cfg.EbayRoleId
	g.ErrorChannelId = cfg.ErrorChannelId
	g.Gumtree = cfg.Gumtree
	g.GumtreeChannelId = cfg.GumtreeChannelId
	g.GumtreeRoleId = cfg.GumtreeRoleId
	g.Salvos = cfg.Salvos
	g.SalvosChannelId = cfg.SalvosChannelId
	g.SalvosRoleId = cfg.SalvosRoleId
	g.SteamQuery = cfg.SteamQuery
	g.SteamQueryChannelId = cfg.SteamQueryChannelId
	g.SteamQueryRoleId = cfg.SteamQueryRoleId

	return q.UpdateGlobals(ctx, g)
}
